### ISSUE TITLE
requirejs/domReady.js can severely delay rendering of content

### ISSUE THREAD

**Issue by DrewML** [magento/magento2#22909]
### Preconditions (*)

1. Have a store running with an optimal Varnish setup (aka be fast to flush the HTML from the server)
2. Use a handler registered with [`requirejs/domReady.js`](https://github.com/magento/magento2/blob/211dd25c73bb2539c3c064374d9d5649f253fd14/lib/web/requirejs/domReady.js) to render some content on the page. As an example, the [Private Content](https://devdocs.magento.com/guides/v2.3/extension-dev-guide/cache/page-caching/private-content.html) feature of Magento works this way, so a default Luma setup is ready to go.

I haven't tested, but I assume this problem does not exist if bundling has been used. When bundling, `domReady.js` will be in a bundle loaded synchronously, so `DOMContentLoaded` will always fire _after_.

### Steps to reproduce (*)

1. In the initial HTML for any page, add a non-blocking resource (async script/img/etc), and make sure the file is served with a severe delay (10+ seconds). `<img src="https://httpstat.us/404?sleep=10000" />` can be used for testing. This is done to simulate real-world delays that can happen with 3rd party content
1. Navigate to the page the 3rd party content (with the simulated delay) has been added to

Note that, after refreshing the page several times, sometimes (race condition) the private content does not render until 10+ seconds have passed.

### Expected result (*)

I expect `domReady.js` to invoke its registered handlers at either of the following times (whichever is observed first):

- When the `DOMContentLoaded` event is fired on the `document`
- When `document.readyState !== 'loading'`

The problem right now is that, if `domReady.js` loads/executes _after_ the `DOMContentLoaded` event has been fired on the `document`, the code falls back to waiting for the `load` event. This is _bad_, and [was a hack for older versions of IE to avoid doing browser sniffing](https://github.com/requirejs/domReady/issues/18).

We should basically borrow [jQuery's implementation of this](https://github.com/jquery/jquery/blob/438b1a3e8a52d3e4efd8aba45498477038849c97/src/core/ready.js#L67-L83), as we know it's supported by all our target browsers. We should _not_ import `jQuery` into `domReady.js`, though. Just copy/paste the implementation (few LOC).

### Actual result (*)

If `domReady.js` loads and executes _after_ `DOMContentLoaded` has already fired, registered handlers will not be invoked until every pending request for a resource has completed. This includes requests for non-critical, 3rd party resources.

**Example**
![image](https://user-images.githubusercontent.com/5233399/57837297-61f88000-7777-11e9-90e6-4083435ab2c8.png)

### What scripts/content get delayed currently by this?
- Any `x-magento-init` or `data-mage-init` widgets (because [`mage/bootstrap.js`](https://github.com/magento/magento2/blob/211dd25c73bb2539c3c064374d9d5649f253fd14/lib/web/mage/bootstrap.js#L9) depends on KnockoutJS, which waits with the `domReady!` plugin)
- [Anything rendering with KnockoutJS](https://github.com/magento/magento2/blob/211dd25c73bb2539c3c064374d9d5649f253fd14/app/code/Magento/Ui/view/base/web/js/lib/knockout/bootstrap.js#L14)
- [Form Key handling](https://github.com/magento/magento2/blob/211dd25c73bb2539c3c064374d9d5649f253fd14/lib/web/mage/common.js#L8)
- [Some responsive functionality on Product/Catalog/Checkout pages](https://github.com/magento/magento2/blob/211dd25c73bb2539c3c064374d9d5649f253fd14/app/design/frontend/Magento/blank/Magento_Theme/web/js/responsive.js#L10)
- [Private Content](https://devdocs.magento.com/guides/v2.3/extension-dev-guide/cache/page-caching/private-content.html)

### Additional Info
This problem is still present in the latest version of requirejs/domReady, [to support browsers that Magento does not support](https://github.com/requirejs/domReady/issues/18).

As `requirejs/domReady.js` is from an older library that really doesn't change much, and is a small amount of code, I'm comfortable with us basically forking the `domReady` loader here and changing the logic inline.

**Comment by DrewML** [comment]
**Placeholder comment where I'll add a video**

**Comment by DrewML** [comment]
@andrewhowdencom [pointed out](https://twitter.com/andrewhowdencom/status/1128946278124195840) on Twitter that I used a bit of a weird example case here.

To replicate, you can also add the following code anywhere in a page:

```html
<img src="https://httpstat.us/404?sleep=10000" />
```

**Comment by kandy** [comment]
Looks like it enough to change
https://github.com/magento/magento2/blob/211dd25c73bb2539c3c064374d9d5649f253fd14/lib/web/requirejs/domReady.js#L92
```
if ( document.readyState !== "loading") {
```
I don't see any other different compare to https://github.com/jquery/jquery/blob/master/src/core/ready.js

**Comment by DrewML** [comment]
@kandy Pretty much!

**Comment by magento-engcom-team** [comment]
@dmanners Thank you for verifying the issue.

Unfortunately, not enough information was provided to acknowledge ticket. Please consider adding the following:
- [ ] Add `"Component: "` label(s) to this ticket based on verification result. If uncertain, you may follow the best guess
- [ ] Add `"Reproduced on "` label(s) to this ticket based on verification result

Once all required information is added, please add label `"Issue: Confirmed"` again.
Thanks!

**Comment by magento-engcom-team** [comment]
@dmanners Thank you for verifying the issue.

Unfortunately, not enough information was provided to acknowledge ticket. Please consider adding the following:
- [ ] Add `"Component: "` label(s) to this ticket based on verification result. If uncertain, you may follow the best guess

Once all required information is added, please add label `"Issue: Confirmed"` again.
Thanks!

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @dmanners 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-17563` were created

**Issue Available**: @dmanners, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by ihor-sviziev** [comment]
This issue wasn't fixed yet, so re-opened

**Comment by magento-engcom-team** [comment]
@ihor-sviziev Thank you for verifying the issue.

Unfortunately, not enough information was provided to acknowledge ticket. Please consider adding the following:
- [x] Add `"Reproduced on "` label(s) to this ticket based on verification result

Once all required information is added, please add label `"Issue: Confirmed"` again.
Thanks!

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @ihor-sviziev 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-17563` were created

**Issue Available**: @ihor-sviziev, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by magento-engcom-team** [comment]
Hi @DrewML, @johnhughes1984. 

Thank you for your report and collaboration!

The issue was fixed by Magento team. The fix was delivered into `magento/magento2:2.3-develop` branch(es).
Related commit(s):
 - [228cb75df6f9804f66aa2377262050c67bfa4ba2](https://github.com/magento/magento2/commit/228cb75df6f9804f66aa2377262050c67bfa4ba2)

The fix will be available with the upcoming `2.3.5` release.

**Comment by krzksz** [comment]
Closing as fixed via https://github.com/magento/magento2/pull/23313
